passes/opt/opt_lut.cpp

```diff
@@ -0,0 +1,268 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2018  whitequark <whitequark@whitequark.org>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/sigtools.h"
+#include "kernel/modtools.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+static bool evaluate_lut(SigMap &sigmap, RTLIL::Cell *lut, dict<SigBit, bool> inputs)
+{
+	SigSpec lut_input = sigmap(lut->getPort("\\A"));
+	int lut_width = lut->getParam("\\WIDTH").as_int();
+	Const lut_table = lut->getParam("\\LUT");
+	int lut_index = 0;
+
+	for (int i = 0; i < lut_width; i++)
+	{
+		SigBit input = sigmap(lut_input[i]);
+		if (inputs.count(input))
+		{
+			lut_index |= inputs[input] << i;
```

Comment:
Will this run into problems if lut_width is greater than the number of bits an in int?

Unclear if that's possible (are these ideal LUTs of any size or device LUTs which will be limited to small widths).

Reply:
Assuming 32-bit ints, that would be a LUT with 4 Gbit of configuration data.

I feel pretty safe.